drivers/serial/uart_nrfx_uart.c

```diff
@@ -242,10 +246,18 @@ static void uart_nrfx_poll_out(struct device *dev,
 	while (!event_txdrdy_check()) {
 	}
 
-	/* Deactivate the transmitter so that it does not needlessly consume
-	 * power.
-	 */
-	nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STOPTX);
+#ifdef CONFIG_UART_0_ASYNC
+	if (uart0_cb.tx_buffer_length == 0) {
+#endif
+		/* Deactivate the transmitter so that it does not needlessly
+		 * consume power.
+		 */
+		nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STOPTX);
```

Comment:
Interrupt driven mode also needs the transmitter to stay active [as long as the TX interrupt is enabled](https://github.com/zephyrproject-rtos/zephyr/pull/22355/files#diff-24c27a01c8713b6b093dfd80f40883d0L745), so it seems that you should also use `device_busy_check(dev)` here (or introduce some flag) to handle that case.